README.md
# raspi_i2c_4deg7seg
Operate 4-digit-7-segment-LED via I2C interface

Target device is 4-Dig.-7Seg. LED module made by HIPOD.
This module can't drive all DP LED.
Both dots of Colon can be lighten in the same time.
The good point of this module is the value .It's not expensive.But it works good.


CHANGELOG.md
## 1.2.1 (12月 04, 2018)
  - Merge branch 'hotfix' into release-1.2  v1.2.0...HEAD
  - Fix version method to answer 1.2 	modified:   ht16k33.py  v1.2.0...HEAD
  - Create new release-1.2 branch  v1.2.0...HEAD
  - Merge from the branch release-1.1  v1.2.0...HEAD
  - Merge from the branch release-1.1  v1.2.0...HEAD
  - Be able to print "FAUL" 	modified:   ht16k33.py  v1.2.0...HEAD
  - Bump version to 1.1.2.  v1.2.0...HEAD
  - Merge from the branch release-1.0  v1.2.0...HEAD
  - Create new release-1.1 branch  v1.2.0...HEAD
  - Merge from the branch release-1.0  v1.2.0...HEAD
  - Add method "version" 	modified:   ht16k33.py  v1.2.0...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v1.0.0.  v1.2.0...HEAD
  - Modify 	modified:   README.md 	modified:   ht16k33.py  v1.2.0...HEAD
  - Modify that if non-nimber letter is passed 	modified:   ht16k33.py  v1.2.0...HEAD
  - Append some 	modified:   README.md  v1.2.0...HEAD
  - Remove 	deleted:    ht16k33_4dig7seg.py 	deleted:    ht16k33_4dig7seg_4dig7seg.py  v1.2.0...HEAD
  - Modify to be verbose in __main__ 	modified:   ht16k33.py  v1.2.0...HEAD
  - Fix 	modified:   ht16k33.py  v1.2.0...HEAD
  - Modulize to be CLASS 	modified:   ht16k33.py  v1.2.0...HEAD
  - Copy form h*4*4* 	new file:   ht16k33.py  v1.2.0...HEAD
  - First commit 	new file:   ht16k33_4dig7seg.py 	new file:   ht16k33_4dig7seg_4dig7seg.py 	new file:   seg.sh  v1.2.0...HEAD
  - Initial commit  v1.2.0...HEAD

## 1.2 (12月 04,2018)
  - Merge from the branch release-1.1  v1.1.2...HEAD
  - Be able to print "FAUL" 	modified:   ht16k33.py  v1.1.2...HEAD
  - Bump version to 1.1.2.  v1.1.2...HEAD
  - Merge from the branch release-1.0  v1.1.2...HEAD
  - Create new release-1.1 branch  v1.1.2...HEAD
  - Add method "version" 	modified:   ht16k33.py  v1.1.2...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v1.0.0.  v1.1.2...HEAD
  - Modify 	modified:   README.md 	modified:   ht16k33.py  v1.1.2...HEAD
  - Modify that if non-nimber letter is passed 	modified:   ht16k33.py  v1.1.2...HEAD
  - Append some 	modified:   README.md  v1.1.2...HEAD
  - Remove 	deleted:    ht16k33_4dig7seg.py 	deleted:    ht16k33_4dig7seg_4dig7seg.py  v1.1.2...HEAD
  - Modify to be verbose in __main__ 	modified:   ht16k33.py  v1.1.2...HEAD
  - Fix 	modified:   ht16k33.py  v1.1.2...HEAD
  - Modulize to be CLASS 	modified:   ht16k33.py  v1.1.2...HEAD
  - Copy form h*4*4* 	new file:   ht16k33.py  v1.1.2...HEAD
  - First commit 	new file:   ht16k33_4dig7seg.py 	new file:   ht16k33_4dig7seg_4dig7seg.py 	new file:   seg.sh  v1.1.2...HEAD
  - Initial commit  v1.1.2...HEAD

## 1.1.2 (12月 04, 2018)
  - Create new release-1.1 branch  v1.1.1...HEAD
  - Merge from the branch release-1.0  v1.1.1...HEAD
  - Add method "version" 	modified:   ht16k33.py  v1.1.1...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v1.0.0.  v1.1.1...HEAD
  - Modify 	modified:   README.md 	modified:   ht16k33.py  v1.1.1...HEAD
  - Modify that if non-nimber letter is passed 	modified:   ht16k33.py  v1.1.1...HEAD
  - Append some 	modified:   README.md  v1.1.1...HEAD
  - Remove 	deleted:    ht16k33_4dig7seg.py 	deleted:    ht16k33_4dig7seg_4dig7seg.py  v1.1.1...HEAD
  - Modify to be verbose in __main__ 	modified:   ht16k33.py  v1.1.1...HEAD
  - Fix 	modified:   ht16k33.py  v1.1.1...HEAD
  - Modulize to be CLASS 	modified:   ht16k33.py  v1.1.1...HEAD
  - Copy form h*4*4* 	new file:   ht16k33.py  v1.1.1...HEAD
  - First commit 	new file:   ht16k33_4dig7seg.py 	new file:   ht16k33_4dig7seg_4dig7seg.py 	new file:   seg.sh  v1.1.1...HEAD
  - Initial commit  v1.1.1...HEAD

## 1.1.1 (12月 04, 2018)
  - Create new release-1.1 branch  v1.1.0...HEAD
  - Merge from the branch release-1.0  v1.1.0...HEAD
  - Add method "version" 	modified:   ht16k33.py  v1.1.0...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v1.0.0.  v1.1.0...HEAD
  - Modify 	modified:   README.md 	modified:   ht16k33.py  v1.1.0...HEAD
  - Modify that if non-nimber letter is passed 	modified:   ht16k33.py  v1.1.0...HEAD
  - Append some 	modified:   README.md  v1.1.0...HEAD
  - Remove 	deleted:    ht16k33_4dig7seg.py 	deleted:    ht16k33_4dig7seg_4dig7seg.py  v1.1.0...HEAD
  - Modify to be verbose in __main__ 	modified:   ht16k33.py  v1.1.0...HEAD
  - Fix 	modified:   ht16k33.py  v1.1.0...HEAD
  - Modulize to be CLASS 	modified:   ht16k33.py  v1.1.0...HEAD
  - Copy form h*4*4* 	new file:   ht16k33.py  v1.1.0...HEAD
  - First commit 	new file:   ht16k33_4dig7seg.py 	new file:   ht16k33_4dig7seg_4dig7seg.py 	new file:   seg.sh  v1.1.0...HEAD
  - Initial commit  v1.1.0...HEAD

## 1.1 (12月 04,2018)
  - Merge from the branch release-1.0  v1.0.0...HEAD
  - Add method "version" 	modified:   ht16k33.py  v1.0.0...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v1.0.0.  v1.0.0...HEAD
  - Modify 	modified:   README.md 	modified:   ht16k33.py  v1.0.0...HEAD
  - Modify that if non-nimber letter is passed 	modified:   ht16k33.py  v1.0.0...HEAD
  - Append some 	modified:   README.md  v1.0.0...HEAD
  - Remove 	deleted:    ht16k33_4dig7seg.py 	deleted:    ht16k33_4dig7seg_4dig7seg.py  v1.0.0...HEAD
  - Modify to be verbose in __main__ 	modified:   ht16k33.py  v1.0.0...HEAD
  - Fix 	modified:   ht16k33.py  v1.0.0...HEAD
  - Modulize to be CLASS 	modified:   ht16k33.py  v1.0.0...HEAD
  - Copy form h*4*4* 	new file:   ht16k33.py  v1.0.0...HEAD
  - First commit 	new file:   ht16k33_4dig7seg.py 	new file:   ht16k33_4dig7seg_4dig7seg.py 	new file:   seg.sh  v1.0.0...HEAD
  - Initial commit  v1.0.0...HEAD

## 1.0.0 (12月 04,2018)




ht16k33_4dig7seg.py
#!/usr/bin/env python3

import smbus

import time
from datetime import datetime

D = [0] * 10
D[0] = 0x3F
D[1] = 0x06
D[2] = 0x5B
D[3] = 0x4F
D[4] = 0x66
D[5] = 0x6D
D[6] = 0x7D
D[7] = 0x27
D[8] = 0x7F
D[9] = 0x6F

DEVICE_BUS = 1
DEVICE_ADDR = 0x70
bus = smbus.SMBus(DEVICE_BUS)
bus.write_byte(DEVICE_ADDR,0x21)
bus.write_byte(DEVICE_ADDR,0x81)
bus.write_byte(DEVICE_ADDR,0xE0)
coron = 0
count = 1 
flag = True
while True:
	try:
		count += 1
		if count == 2:
			ms = datetime.now().strftime("%M%S")
			d1 = int(ms[0])
			d2 = int(ms[1])
			d3 = int(ms[2])
			d4 = int(ms[3])
			count = 0
		if flag:
			coron = 2
			flag = False
		else:
			coron = 0
			flag = True
		data = [D[d1],0,D[d2],0,coron,0,D[d3],0,D[d4]]
		bus.write_i2c_block_data(DEVICE_ADDR,0x00,data)
		time.sleep(0.5)

	except KeyboardInterrupt:
		break



ht16k33_4dig7seg_4dig7seg.py
#!/usr/bin/env python3

#https://www.javadrive.jp/python/string/index9.html
#数値を文字列に変換(str)




import smbus

import time
from datetime import datetime

D = [0] * 10
D[0] = 0x3F
D[1] = 0x06
D[2] = 0x5B
D[3] = 0x4F
D[4] = 0x66
D[5] = 0x6D
D[6] = 0x7D
D[7] = 0x27
D[8] = 0x7F
D[9] = 0x6F

DEVICE_BUS = 1
DEVICE_ADDR0 = 0x70
DEVICE_ADDR1 = 0x71
bus = smbus.SMBus(DEVICE_BUS)
bus.write_byte(DEVICE_ADDR0,0x21)
bus.write_byte(DEVICE_ADDR0,0x81)
bus.write_byte(DEVICE_ADDR0,0xE0)
bus.write_byte(DEVICE_ADDR1,0x21)
bus.write_byte(DEVICE_ADDR1,0x81)
bus.write_byte(DEVICE_ADDR1,0xE0)
coron = 0
count = 1 
flag = True
while True:
	try:
		count += 1
		if count == 2:
			m = datetime.now().strftime("%M")
			s = datetime.now().strftime("%S")
			d01 = int(m[0])
			d02 = int(m[1])
			d03 = int(s[0])
			d04 = int(s[1])
			a = "0000"+str(int(m) * int(s))
			d11 = int(a[-4])
			d12 = int(a[-3])
			d13 = int(a[-2])
			d14 = int(a[-1])

			count = 0
		if flag:
			coron = 2
			flag = False
		else:
			coron = 0
			flag = True
		data0 = [D[d01],0,D[d02],0,coron,0,D[d03],0,D[d04]]
		data1 = [D[d11],0,D[d12],0,0,0,D[d13],0,D[d14]]
		bus.write_i2c_block_data(DEVICE_ADDR0,0x00,data0)
		bus.write_i2c_block_data(DEVICE_ADDR1,0x00,data1)

		time.sleep(0.5)

	except KeyboardInterrupt:
		break



ht16k33.py
#!/usr/bin/env python3

#https://www.javadrive.jp/python/string/index9.html
#数値を文字列に変換(str)


import smbus

import time

class Ht16k33:
	DDD = { }
	DDD["0"] = 0x3F
	DDD["1"] = 0x06
	DDD["2"] = 0x5B
	DDD["3"] = 0x4F
	DDD["4"] = 0x66
	DDD["5"] = 0x6D
	DDD["6"] = 0x7D
	DDD["7"] = 0x27
	DDD["8"] = 0x7F
	DDD["9"] = 0x6F
	DDD["-"] = 0x00
	DDD["F"] = 0x71
	DDD["A"] = 0x77
	DDD["U"] = 0x3E
	DDD["L"] = 0x38

	def __init__(self,dev_bus,dev_addr):
		#(hex) dev_bus
		#(hex) dev_addr
		self.bus = smbus.SMBus(dev_bus)
		self.addr = dev_addr
		self.bus.write_byte(self.addr,0x21)
		self.bus.write_byte(self.addr,0x81)
		self.bus.write_byte(self.addr,0xE0)

	def version(self):
		return "1.2"


		#light strength
	def strength(self,value):
		#(hex) value : 0x0～0xF
		value = int(value)
		if value > 0xF:
			value = 0xF
		elif value < 0x0:
			value = 0x0
		value = value + 0xE0;
		self.bus.write_byte(self.addr,value)

	def print(self,sValue="0000",fColon=False):
		#(String) sValue
		#(boolean) fColon : 0:disappear  1:appear
		sValue = "----" + sValue
		flag = True
		try:
			d4 = Ht16k33.DDD[sValue[-1]]
			d3 = Ht16k33.DDD[sValue[-2]]
			d2 = Ht16k33.DDD[sValue[-3]]
			d1 = Ht16k33.DDD[sValue[-4]]
		except:
			print("sValue contains another of [0-9] letter")
			flag = False
		if flag:
			if fColon == True:
				coron = 0xf
			else:
				coron = 0x0
			data = [d1,0,d2,0,coron,0,d3,0,d4]
			self.bus.write_i2c_block_data(self.addr,0x00,data)

	def close(self):
		print("you don't have to close smbus")

if __name__ == "__main__":

	#SAMPLE
	#	usage:
	#		example1 : $ python3 ht16k33.py 1
	#			if you connect 1 4Dig7Sed-LED
	#		example2 : $ python3 ht16k33.py 2
	#			if you connect 2 4Dig7Sed-LEDs

	from datetime import datetime
	import sys
	
	arg = sys.argv
	if len(arg) == 1:
		print("enter number of 4dig7seg-LED device(s),expect [1-2]")
		h = Ht16k33(1,0x70)
		h.print("FAUL",2)
	else:
		if arg[1] == "1":
			flag1 = True
			flag2 = False
		elif arg[1] == "2":
			flag1 = True
			flag2 = True
		else:
			print("Wrong number,expect [1-2]")
			flag1 = False
			flag2 = False

		if flag1:
			DEVICE_BUS = 1
			DEVICE_ADDR0 = 0x70
			DEVICE_ADDR1 = 0x71

			h1 = Ht16k33(DEVICE_BUS,DEVICE_ADDR0)
			if flag2:h2 = Ht16k33(DEVICE_BUS,DEVICE_ADDR1)

			coron = 0
			count = 1 
			flag = True
			while True:
				try:
					count += 1
					if count == 2:
						ms = datetime.now().strftime("%M%S")
						mult = datetime.now().strftime("%d%H")

						count = 0
					if flag:
						flag = False
						coron = True
					else:
						flag = True
						coron = False
					h1.print(ms,coron)
					if flag2:h2.print(mult,coron)

					time.sleep(0.5)

				except KeyboardInterrupt:
					break

